File: checker_board.py
import copy


class checkerBoard:
    """Makes a standard 8x8 checker board"""
    def __init__(self):
        self.board_size = 8  # standard checker board size
        self.blank_piece = {"red": False, "black": False, "king": False}
        self.red_turn = True
        self.make_board()
        self.setup_board()
        self.movable_pieces = self.get_movable_pieces()

    def make_board(self):
        self.board = []

        for row in range(self.board_size):
            self.board.append([])
            for column in range(self.board_size):
                self.board[row].append(copy.deepcopy(self.blank_piece))

    def setup_board(self, rows_per_side=3):
        offset = False
        for row in range(rows_per_side):
            offset = not offset
            for column in range(0, self.board_size, 2):
                self.board[row][column+offset]["black"] = True
        self.board.reverse()

        offset = True
        for row in range(rows_per_side):
            offset = not offset
            for column in range(0, self.board_size, 2):
                self.board[row][column+offset]["red"] = True
        self.board.reverse()

    def print_board(self):
        row_num = 0
        print("———0———1———2———3———4———5———6———7——")
        for row in self.board:

            line = f"{row_num}| "
            row_num += 1

            for column in row:

                if column["red"]:
                    line += "R | "
                elif column["black"]:
                    line += "B | "
                else:
                    line += "  | "

            print(line)
            print("——————————————————————————————————")

    def get_movable_pieces(self):
        """returns a list of the locations of pieces that can be moved"""
        movable_pieces = []

        for piece in self.get_player_pieces():
            row, column = piece[0], piece[1]
            
            if self.board[row][column]["king"]:
                movable = (self.__check_empty_above__(row, column) or
                           self.__check_empty_below__(row, column) or
                           self.__check_capture_above__(row, column) or
                           self.__check_capture_below__(row, column))
            elif self.board[row][column]["black"]:
                movable = (self.__check_empty_below__(row, column) or
                           self.__check_capture_below__(row, column))
            else:
                movable = (self.__check_empty_above__(row, column) or
                           self.__check_capture_above__(row, column))
            if movable:
                movable_pieces.append(piece)
        return movable_pieces

    def move_piece(self, row, column, destination_row, destination_column):
        """Will move the given piece to the destination if it is a legal move"""
        complete_move = False
        captured = False
        if [row, column] in self.movable_pieces:
            if destination_row > row and (self.board[row][column]["black"] or
                                          self.board[row][column]["king"]):
                if abs(destination_column - column) == 1 and [destination_row, destination_column] in self.__get_empty_below__(row, column):
                    complete_move = True
                elif abs(destination_column - column) == 2:
                    possible_end_spaces, captures = self.__get_capture_below__(row, column)
                    if [destination_row, destination_column] in possible_end_spaces:
                        if destination_column > column:
                            self.board[row+1][column+1] = copy.deepcopy(self.blank_piece)
                            complete_move = True
                            captured = True
                        else:
                            self.board[row+1][column-1] = copy.deepcopy(self.blank_piece)
                            complete_move = True
                            captured = True
            if destination_row < row and (self.board[row][column]["red"] or
                                          self.board[row][column]["king"]):
                if abs(destination_column - column) == 1 and [destination_row, destination_column] in self.__get_empty_above__(row, column):
                    complete_move = True
                elif abs(destination_column - column) == 2:
                    possible_end_spaces, captures = self.__get_capture_above__(row, column)
                    if [destination_row, destination_column] in possible_end_spaces:
                        if destination_column > column:
                            self.board[row-1][column+1] = copy.deepcopy(self.blank_piece)
                            complete_move = True
                            captured = True
                        else:
                            self.board[row-1][column-1] = copy.deepcopy(self.blank_piece)
                            complete_move = True
                            captured = True

        
        if complete_move:
            self.board[destination_row][destination_column] = copy.deepcopy(self.board[row][column])
            self.board[row][column] = copy.deepcopy(self.blank_piece)
            if not captured:
                self.red_turn = not self.red_turn
                self.movable_pieces = self.get_movable_pieces()
        
        if captured:
            if (self.board[destination_row][destination_column]["black"] or self.board[destination_row][destination_column]["king"]) and self.__check_capture_below__(destination_row, destination_column):
                self.movable_pieces = [[destination_row, destination_column]]
            elif (self.board[destination_row][destination_column]["red"] or self.board[destination_row][destination_column]["king"]) and self.__check_capture_above__(destination_row, destination_column): 
                self.movable_pieces = [[destination_row, destination_column]]
            else:
                self.red_turn = not self.red_turn
                self.movable_pieces = self.get_movable_pieces()

    def get_all_pieces(self):
        possible_pieces = []
        for row in range(self.board_size):
            for column in range(self.board_size):
                if (self.board[row][column]["black"] or self.board[row][column]["red"]):
                    possible_pieces.append([row, column])

        return possible_pieces

    def get_player_pieces(self):
        possible_pieces = []
        for row in range(self.board_size):
            for column in range(self.board_size):
                if self.board[row][column]["red"] and self.red_turn:
                    possible_pieces.append([row, column])
                elif self.board[row][column]["black"] and not self.red_turn:
                    possible_pieces.append([row, column])

        return possible_pieces

    def __check_capture_above__(self, row, column):
        """Checks above a given piece if there is any space to move to"""
        can_capture = False
        if self.board[row][column]["red"]:
            if column < self.board_size-2 and row > 1:
                if self.board[row-1][column+1]["black"] and (
                        not self.board[row-2][column+2]["red"] and
                        not self.board[row-2][column+2]["black"]):
                    can_capture = True
            elif column > 1 and row > 1:
                if self.board[row-1][column-1]["black"] and (
                        not self.board[row-2][column-2]["red"] and
                        not self.board[row-2][column-2]["black"]):
                    can_capture = True

        elif self.board[row][column]["black"]:
            if column < self.board_size-2 and row > 1:
                if self.board[row-1][column+1]["red"] and (
                        not self.board[row-2][column+2]["red"] and
                        not self.board[row-2][column+2]["black"]):
                    can_capture = True
            elif column > 1 and row > 1:
                if self.board[row-1][column-1]["red"] and (
                        not self.board[row-2][column-2]["red"] and
                        not self.board[row-2][column-2]["black"]):
                    can_capture = True

        return can_capture

    def __check_capture_below__(self, row, column):
        """Checks above a given piece if there is any space to move to"""
        can_capture = False
        if self.board[row][column]["red"]:
            if column < self.board_size-2 and row < self.board_size-2:
                if self.board[row+1][column+1]["black"] and (
                        not self.board[row+2][column+2]["red"] and
                        not self.board[row+2][column+2]["black"]):
                    can_capture = True
            elif column > 1 and row < self.board_size-2:
                if self.board[row+1][column-1]["black"] and (
                        not self.board[row+2][column-2]["red"] and
                        not self.board[row-2][column-2]["black"]):
                    can_capture = True

        elif self.board[row][column]["black"]:
            if column < self.board_size-2 and row < self.board_size-2:
                if self.board[row+1][column+1]["red"] and (
                        not self.board[row+2][column+2]["red"] and
                        not self.board[row+2][column+2]["black"]):
                    can_capture = True
            elif column > 1 and row < self.board_size-2:
                if self.board[row+1][column-1]["red"] and (
                        not self.board[row+2][column-2]["red"] and
                        not self.board[row+2][column-2]["black"]):
                    can_capture = True

        return can_capture

    def __check_empty_above__(self, row, column):
        """Checks above a given piece if there is any space to move to"""
        movable = False
        if column < self.board_size-1 and row > 0:
            if (not self.board[row-1][column+1]["red"] and
                    not self.board[row-1][column+1]["black"]):
                movable = True
        if column > 0 and row > 0:
            if (not self.board[row-1][column-1]["red"] and
                    not self.board[row-1][column-1]["black"]):
                movable = True

        return movable

    def __check_empty_below__(self, row, column):
        """Checks above a given piece if there is any space to move to"""
        movable = False
        if column < self.board_size-1 and row < self.board_size-1:
            if (not self.board[row+1][column+1]["red"] and
                    not self.board[row+1][column+1]["black"]):
                movable = True
        if column > 0 and row < self.board_size-1:
            if (not self.board[row+1][column-1]["red"] and
                    not self.board[row+1][column-1]["black"]):
                movable = True

        return movable

    def __get_capture_above__(self, row, column):
        """Checks above a given piece if there is any space to move to"""
        end_spaces, captures = [], []
        if self.board[row][column]["red"]:
            if column < self.board_size-2 and row > 1:
                if self.board[row-1][column+1]["black"] and (
                        not self.board[row-2][column+2]["red"] and
                        not self.board[row-2][column+2]["black"]):
                    captures.append([row-1, column+1])
                    end_spaces.append([row-2, column+2])
            elif column > 1 and row > 1:
                if self.board[row-1][column-1]["black"] and (
                        not self.board[row-2][column-2]["red"] and
                        not self.board[row-2][column-2]["black"]):
                    captures.append([row-1, column-1])
                    end_spaces.append([row-2, column-2])

        elif self.board[row][column]["black"]:
            if column < self.board_size-2 and row > 1:
                if self.board[row-1][column+1]["red"] and (
                        not self.board[row-2][column+2]["red"] and
                        not self.board[row-2][column+2]["black"]):
                    captures.append([row-1, column+1])
                    end_spaces.append([row-2, column+2])
            elif column > 1 and row > 1:
                if self.board[row-1][column-1]["red"] and (
                        not self.board[row-2][column-2]["red"] and
                        not self.board[row-2][column-2]["black"]):
                    captures.append([row-1, column-1])
                    end_spaces.append([row-2, column-2])

        return end_spaces, captures

    def __get_capture_below__(self, row, column):
        """Checks above a given piece if there is any space to move to"""
        end_spaces, captures = [], []
        if self.board[row][column]["red"]:
            if column < self.board_size-2 and row < self.board_size-2:
                if self.board[row+1][column+1]["black"] and (
                        not self.board[row+2][column+2]["red"] and
                        not self.board[row+2][column+2]["black"]):
                    captures.append([row+1, column+1])
                    end_spaces.append([row+2, column+2])
            elif column > 1 and row < self.board_size-2:
                if self.board[row+1][column-1]["black"] and (
                        not self.board[row+2][column-2]["red"] and
                        not self.board[row-2][column-2]["black"]):
                    captures.append([row+1, column-1])
                    end_spaces.append([row+2, column-2])

        elif self.board[row][column]["black"]:
            if column < self.board_size-2 and row < self.board_size-2:
                if self.board[row+1][column+1]["red"] and (
                        not self.board[row+2][column+2]["red"] and
                        not self.board[row+2][column+2]["black"]):
                    captures.append([row+1, column+1])
                    end_spaces.append([row+2, column+2])
            elif column > 1 and row < self.board_size-2:
                if self.board[row+1][column-1]["red"] and (
                        not self.board[row+2][column-2]["red"] and
                        not self.board[row+2][column-2]["black"]):
                    captures.append([row+1, column-1])
                    end_spaces.append([row+2, column-2])

        return end_spaces, captures

    def __get_empty_above__(self, row, column):
        """returns empty spaces that the given piece can move to"""
        spaces = []
        if column < self.board_size-1 and row > 0:
            if (not self.board[row-1][column+1]["red"] and
                    not self.board[row-1][column+1]["black"]):
                spaces.append([row-1, column+1])
        if column > 0 and row > 0:
            if (not self.board[row-1][column-1]["red"] and
                    not self.board[row-1][column-1]["black"]):
                spaces.append([row-1, column-1])

        return spaces

    def __get_empty_below__(self, row, column):
        """returns empty spaces that the given piece can move to"""
        spaces = []
        if column < self.board_size-1 and row < self.board_size-1:
            if (not self.board[row+1][column+1]["red"] and
                    not self.board[row+1][column+1]["black"]):
                spaces.append([row+1, column+1])
        if column > 0 and row < self.board_size-1:
            if (not self.board[row+1][column-1]["red"] and
                    not self.board[row+1][column-1]["black"]):
                spaces.append([row+1, column-1])

        return spaces


if __name__ == "__main__":
    game = checkerBoard()

    while True:
        if game.red_turn:
            print("red Turn")
        else:
            print("black Turn")
        game.print_board()
        
        game.move_piece(5, 0, 4, 1)
        print([4, 1] in game.__get_empty_above__(5, 0))
        print(input("yay"))

File: checkerBoard.lua
dofile("checkerPiece.lua")
dofile("move.lua")
dofile("helperFunctions.lua")


---@class board
---@field playarea baseCheckerPiece[][] 2d array representing all the spots on a board
---@field redTurn boolean it will always be someone's turn
---@field redWin boolean
---@field blackWin boolean
---@field redCaptured integer count of all red pieces that were captured
---@field blackCaptured integer count of all black pieces that were captured
---@field turns integer how many turns have been taken
---@field width integer width of the playarea measured in tiles
---@field height integer height of the playarea measured in tiles
board = {}

---@param width integer | nil
---@param height integer | nil
---@return board
function board.init(width, height)
    local self = table.deepCopy(board)
    ---standard checker board size is 8x8
    self.playarea = {}

    self.width = width or 8
    self.height = height or 8
    self.redTurn = true
    self.redWin = false
    self.blackWin = false
    self.redCaptured = 0
    self.blackCaptured = 0
    self.turns = 0

    ---creates board
    for i = 1, self.height, 1 do
        local temp = {}
        for j = 1, self.width, 1 do
            table.insert(temp, {})
        end
        table.insert(self.playarea, temp)
    end


    return self
end

---will print the current board to the console
---@param printTurn boolean
---@param showPieceMoves boolean
function board:printBoard(printTurn, showPieceMoves)
    if printTurn then
        if self.redTurn then
            print("red turn")
        else
            print("black turn")
        end
    end

    local rowNum = 1
    local head = ""
    local lineSep = ""
    for i = 1, self.width, 1 do
        head = head .. "———" .. tostring(i)
        lineSep = lineSep .. "————"
    end

    head = head .. "——"
    lineSep = lineSep .. "——"
    print(head)
    for i, row in ipairs(self.playarea) do
        local line = tostring(rowNum) .. "| "
        rowNum = rowNum + 1
        for j, square in ipairs(row) do
            if square.isRed == nil then -- will be nil since it is undefined in blank spots
                line = line .. "  | "
            else
                local possibleMoves = #square:getPossibleMoves(self)
                if square.pieceName == "king" then
                    line = line .. "\27[1m"
                end
                if square.isRed then
                    line = line .. "\27[31m"
                else
                    line = line .. "\27[34m"
                end

                if showPieceMoves then
                    line = line .. possibleMoves
                else
                    if square.isRed then
                        line = line .. "R"
                    else
                        line = line .. "B"
                    end
                end

                line = line .. "\27[0m | "
            end
        end
        print(line)
        print(lineSep)
    end
end

function board:terminalGame()
    self:setupCheckers()

    while true do
        self:printBoard(true, true)
        print("Choose a piece")
        print("row:")
        local row = tonumber(io.read())
        print("column:")
        local column = tonumber(io.read())
        if self.playarea[row][column].isRed == nil then
            print("Not a piece!")
        else
            print("Choose a destination")
            print("row:")
            local drow = tonumber(io.read(), 10)
            print("column:")
            local dcolumn = tonumber(io.read(), 10)
            if not self:makeMove(self.playarea[row][column].position, coordinate.init(drow, dcolumn)) then
                print("\27[1m\27[31mMOVE NOT POSSIBLE!\27[0m")
            end
        end
    end
end

---populates the playarea with checker pieces arranged in a standard fassion
---@param rowsPerSide integer | nil number of rows to populate per side defaults to 3
function board:setupCheckers(rowsPerSide)
    rowsPerSide = rowsPerSide or 3
    local offset = false
    for row = 1, rowsPerSide, 1 do
        offset = not offset
        for column = 1, self.width, 2 do
            if offset then
                self.playarea[row][column + 1] = baseCheckerPiece.init(false, coordinate.init(row, column + 1))
            else
                self.playarea[row][column] = baseCheckerPiece.init(false, coordinate.init(row, column))
            end
        end
    end

    offset = true
    for row = self.height - rowsPerSide + 1, self.height, 1 do
        offset = not offset
        for column = 1, self.width, 2 do
            if offset then
                self.playarea[row][column + 1] = baseCheckerPiece.init(true, coordinate.init(row, column + 1))
            else
                self.playarea[row][column] = baseCheckerPiece.init(true, coordinate.init(row, column))
            end
        end
    end
end

---attempt a move
---@param start coordinate
---@param destination coordinate
---@return boolean
function board:makeMove(start, destination)
    local pieceIsRed = self.playarea[start.row][start.column].isRed
    ---checkes if there is a piece to move
    if pieceIsRed == nil then
        return false
    end

    ---checkes if it is the proper turn
    if self.redTurn ~= pieceIsRed then
        return false
    end

    ---makes sure only pieces marked as mustMove can be moved
    if self:mustMoves(self.redTurn) and not self.playarea[start.row][start.column].mustMove then
        return false
    end

    local moveSuccess = self.playarea[start.row][start.column]:move(self, destination)
    if moveSuccess and not self.playarea[destination.row][destination.column].mustMove then
        self.redTurn = not self.redTurn
    end
    return moveSuccess
end

---checks the board for pieces that are marked as mustMove
---@param red boolean if true checks for red mustMoves; false checks for black
---@return boolean
function board:mustMoves(red)
    for _, row in pairs(self.playarea) do
        for key, square in pairs(row) do
            if square.isRed == red and square.mustMove then
                return true
            end
        end
    end
    return false
end

--- game = board.init()
--- game:terminalGame()
